### ISSUE TITLE
ACME S/MIME support

### ISSUE THREAD

**Issue by LecrisUT** [smallstep/certificates#730]
I have just found that ACME now support S/MIME via [rfc8823](https://datatracker.ietf.org/doc/rfc8823/). While the implementations of this is still ongoing, there is already a client being developed ontop of certbot: [polhenarejos/acme_email](https://github.com/polhenarejos/acme_email).

What are your thoughts on this?

**Comment by LecrisUT** [comment]
If both parties have different CAs then yes, it would be less useful. But if they have different domains but connect to the same CA, you get the benefits of S/MIME, signature verification and/or encryption. Probably also useful in organization setups, with different departments.

I'm still looking into this, but if it's possible to combine with dmarc and publish the root CA via dnssec, that would be something. Hopefully though let's encrypt will implement this acme to be the public signature.

**Comment by hslatman** [comment]
I think you can find S/MIME deployments with internal PKI in (large) Windows deployments with elevated (internal) email security requirements. You'll likely find AD CS and perhaps also SCEP in use there, so then that would've to be (partly) replaced by step. 

Something that I just thought of: assuming that it is possible to integrate with public CAs using their (proprietary) interfaces, it may be possible to function like a gateway to that API, verifying the requestor identity and handling the ACME S/MIME challenge in step, resulting in an S/MIME certificate from that public CA. Conceptually it is similar to the cloudcas integration.

**Comment by mmalone** [comment]
Yea, it seems like root distribution is the problem here. I can understand the elevated internal security use case, but I don't know how common that is. If there's a way to do root distribution (e.g., with DMARC? I didn't know DMARC did that though?) then this becomes more interesting.

It _feels like_ the main value of RFC8823 is to make it easier for end-users to get publicly trusted S/MIME certificates. but I definitely could be wrong. I can see value in implementing this spec just to have an open source reference implementation for people to test against and whatnot, but that's fairly niche, and I think this could be a fair amount of work. The RFC itself looks pretty simple, but we don't currently send or receive email from `step-ca`, so we'll have to figure that out.

For now, let's leave this open. **If anyone else from the community wants this, please express interest by giving this issue a 👍 or, even better, leaving a comment with a couple sentences on why you need this.**

**Comment by LecrisUT** [comment]
> (e.g., with DMARC? I didn't know DMARC did that though?)

I did more digging on the DMARC part, and it seems there is no support for anything other than SPF and DKIM. I guess it's because it's supposed to be domain wide policy, while S/MIME is for the end-user.

For now indeed, the ACME part of S/MIME is only useful if you can get the root to be public or you have a big enough cooperation.

There is the certificate DNS record if you have DNSSEC on, but I don't know how that one interacts in general.

As for S/MIME, a second use-case is for signing commits. Useful if you have something like an internal gitea repository. But if you also want to make it public, it could become problematic.

For now at least, how about making a template for S/MIME through the current providers? Do the current client certificates already work for that?

**Comment by mmalone** [comment]
Yes, you should be able to use templates to issue certificates with the right extensions (key usage / extended key usage) for S/MIME using any of our existing provisions (OIDC would work really well for this). We don't include the necessary extensions by default though. You may also need to use RSA keys since S/MIME also does encryption (you can't encrypt with ECDSA keys).

For code signing we're keeping a close eye on https://www.sigstore.dev/.

**Comment by jwildeboer** [comment]
FTR: I have gone through the work of using step-ca to create (manually) working S/MIME certs and documented it all at https://jan.wildeboer.net/2025/08/Create-SMIME-Cert-stepca/ 

I imported the generated certificate into Mail on MacOS and iOS and they work perfectly for signing and encrypting, yes, with ECDSA. We also cross-checked with Microsoft Outlook, Thunderbird and Evolution. They all support this.

No big surprise, as the current baseline requirements from the CA/Browser Forum demand support for ECDSA and actually advice against using RSA anymore.

https://cabforum.org/working-groups/smime/requirements/

At https://acme.castle.cloud you can also find a working implementation of the  Email-Reply-00 challenge.

**Comment by hslatman** [comment]
> FTR: I have gone through the work of using step-ca to create (manually) working S/MIME certs and documented it all at https://jan.wildeboer.net/2025/08/Create-SMIME-Cert-stepca/

Nice blog 🙂 

> I imported the generated certificate into Mail on MacOS and iOS and they work perfectly for signing and encrypting, yes, with ECDSA. We also cross-checked with Microsoft Outlook, Thunderbird and Evolution. They all support this.

Do you have more details about this? Does it use ECDH or ECIES? I suppose the content is encrypted using symmetric encryption, with a key derived using ECDH, in line with PKCS7/CMS. Parts of this could be learned from the email PEM.

> No big surprise, as the current baseline requirements from the CA/Browser Forum demand support for ECDSA and actually advice against using RSA anymore.
> 
> https://cabforum.org/working-groups/smime/requirements/
> 
> At https://acme.castle.cloud you can also find a working implementation of the Email-Reply-00 challenge.

I worked on an ACME client to add/fix support for it: https://github.com/mholt/acmez/pull/25, and validated it against that ACME server. I've also created a POC using `step-ca` with support that also works with said client.

**Comment by jwildeboer** [comment]
If you send me a signed ping email at jan [at] wildeboer DOT net, I can reply so you have my cert and we can exchange some encrypted mails and I can send you raw mails for analysis.
